contracts/two-party-pol-holder/src/msg.rs

```diff
@@ -0,0 +1,233 @@
+            Attribute::new("clock_addr", self.clock_address),
+            Attribute::new("pool_address", self.pool_address),
+            Attribute::new("next_contract", self.next_contract),
+        ];
+        attrs.extend(self.ragequit_config.get_response_attributes());
+        attrs.extend(self.lockup_config.get_response_attributes());
+        attrs
+    }
+}
+
+#[cw_serde]
+pub struct TwoPartyPolCovenantConfig {
+    pub party_a: TwoPartyPolCovenantParty,
+    pub party_b: TwoPartyPolCovenantParty,
+}
+
+impl TwoPartyPolCovenantConfig {
+    pub fn update_parties(&mut self, p1: TwoPartyPolCovenantParty, p2: TwoPartyPolCovenantParty) {
+        if self.party_a.addr == p1.addr {
+            self.party_a = p1;
+            self.party_b = p2;
+        } else {
+            self.party_a = p2;
+            self.party_b = p1;
+        }
+    }
+
+    pub fn validate(&self, api: &dyn Api) -> Result<(), ContractError> {
+        api.addr_validate(&self.party_a.router)?;
+        api.addr_validate(&self.party_b.router)?;
+        if self.party_a.allocation + self.party_b.allocation != Decimal::one() {
+            return Err(ContractError::AllocationValidationError {});
+        }
+        Ok(())
+    }
+}
+
+#[cw_serde]
+pub struct TwoPartyPolCovenantParty {
+    /// the `denom` and `amount` (`Uint128`) to be contributed by the party
+    pub contribution: Coin,
+    /// address authorized by the party to perform claims/ragequits
+    pub addr: String,
+    /// fraction of the entire LP position owned by the party.
+    /// upon exiting it becomes 0.00. if counterparty exits, this would
+    /// become 1.00, meaning that this party owns the entire position
+    /// managed by the covenant.
+    pub allocation: Decimal,
+    /// address of the interchain router associated with this party
+    pub router: String,
+}
+
+impl TwoPartyPolCovenantConfig {
+    /// if authorized, returns (party, counterparty). otherwise errors
+    pub fn authorize_sender(
+        &self,
+        sender: &Addr,
+    ) -> Result<(TwoPartyPolCovenantParty, TwoPartyPolCovenantParty), ContractError> {
+        let party_a = self.party_a.clone();
+        let party_b = self.party_b.clone();
+        if party_a.addr == *sender {
+            Ok((party_a, party_b))
+        } else if party_b.addr == *sender {
+            Ok((party_b, party_a))
+        } else {
+            Err(ContractError::Unauthorized {})
+        }
+    }
+}
+
+#[clocked]
+#[cw_serde]
+pub enum ExecuteMsg {
+    /// initiate the ragequit
+    Ragequit {},
+    /// withdraw the liquidity party is entitled to
+    Claim {},
+}
+
+#[cw_serde]
+pub enum ContractState {
+    /// contract is instantiated and awaiting for deposits from
+    /// both parties involved
+    Instantiated,
+    /// funds have been forwarded to the LP module. from the perspective
+    /// of this contract that indicates an active LP position.
+    /// TODO: think about whether this is a fair assumption to make.
```

Comment:
Technically we can't assume this because it's possible there are non LP'd funds in the LPer, depending on the LP config. We need to account for how to distribute these funds when parties exit/rage quit

Reply:
I see your point, but I see any LP contract issues (like leftover funds) as an issue of the LP, not holder. Holder did complete its task of routing the funds to the next module..

The simplest fix is probably to add a permisionless transfer method on the LP contract that would route all its balances to the holder. Splitter fallback config could be used to distribute those, but given that we want to get rid of it, I'll see how to add that logic  to the holder itself.